crates/sui-framework/sources/display.move

```diff
@@ -0,0 +1,184 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Defines a Display struct which defines the way an Object
+/// should be displayed. The intention is to keep data as independent
+/// from its display as possible, protecting the development process
+/// and keeping it separate from the ecosystem agreements.
+///
+/// Each of the fields of the Display object should allow for pattern
+/// substitution and filling-in the pieces using the data from the object T.
+///
+/// More entry functions might be added in the future depending on the use cases.
+module sui::display {
+    use sui::publisher::{is_package, Publisher};
+    use sui::tx_context::TxContext;
+    use std::string::{String, utf8};
+    use sui::vec_map::{Self, VecMap};
+    use sui::object::{Self, ID, UID};
+    use sui::transfer;
+    use sui::event;
+    use std::vector;
+
+    /// For when T does not belong to package in PublisherCap.
+    const ENotOwner: u64 = 0;
+
+    /// For when vectors passed into one of the multiple insert functions
+    /// don't match in their lengths.
+    const EVecLengthMismatch: u64 = 1;
+
+    /// The Display object. Defines the way an object should be
+    /// displayed. Display object can only be created and modified with
+    /// a PublisherCap, making sure that the rules are set by the owner
+    /// of the type.
+    ///
+    /// Each of the display properties should support patterns outside
+    /// of the system, making it simpler to customize Display based
+    /// on the property values of an Object.
+    /// ```
+    /// // Example of a display object
+    /// Display<0x...::capy::Capy> {
+    ///  fields:
+    ///    <name, "Capy {{ genes }}">
+    ///    <link, "https://capy.art/capy/{{ id }}">
+    ///    <image, "https://api.capy.art/capy/{{ id }}/svg">
+    ///    <description, "Lovely Capy, one of many">
+    /// }
+    /// ```
+    ///
+    /// Uses only String type due to external-facing nature of the object,
+    /// the property names have a priority over their types.
+    struct Display<phantom T: key> has key {
+        id: UID,
+        /// Contains fields for display. Currently supported
+        /// fields are: name, link, image and description.
+        fields: VecMap<String, String>,
```

Comment:
Might make sense to have a `version` that increments each time the fields are updated (or maybe the object version is enough here, not sure...).

Basically, I'm wondering how a client will deal with a change in `Display`, but an explicit version field here may or may not be part of that story.

Reply:
+1 with the ability to arbitrarily add/remove fields it might be a bit tricky to track changes overtime? 

Or do we just plan on relying on the Object's Version?

Reply:
Would it make sense to have something like a `format` field in `Display`? Not in the "object version" sense described above, but rather to indicate the display standard to which a `Display` object fields conform. Something like `format='sui-1.0.0'`, `format='alternativeStandard-1.0.0'`, etc.

Maybe the presence/absence of certain fields will be enough. But I wonder if there might be some scenario later down the line where this could be needed - for example if multiple display standards emerge and they have overlapping field names.

Reply:
@sblackshear @tnowacki absolutely agree that version tracking should be a part of the Display standard. However, for moving forward I would ignore it for now and get back to it once we understand the need for versioning deeper and study the use cases.

The version won't be left intact but I'd rather see the Display as a part of the framework sooner and start testing the basics and implement support for versioning later. Though it's worth keeping it in mind while designing the RPC endpoints to allow passing version in the future.

Reply:
I would suggest we go ahead and add the field, but leave the logic for auto-updating it out (just expose a setter for the `Publisher` to change if/when they wish) + skip support in the SDK or elsewhere for now. That will let us get going quickly without getting caught in an un-versionable trap later.